### PR TITLE
Add spec tests for HTTP-Trace 

### PR DESCRIPTION
**Pull request by 3V3RYONE**
Created a new method in **Metasploit::Framework::LoginScanner::HTTP#send_http_trace_proc()** which defines the proc related to HTTP-Trace. Earlier, the proc was defined in the **#send_request()** method. The **#send_request()** method works upon multiple things like crafting a HTTP request, sending the request to the server and receiving the response.  
  
Thus, it is difficult to test the HTTP-Trace proc in the vast **#send_request()** method. But the new method (**#set_http_trace_proc()**) just deals with defining the proc, thus making the code readable, modular and easier to test.  
  
Added the Rspec unit tests to check the behavior of the following methods:  
1. **Metasploit::Framework::LoginScanner::HTTP#set_http_trace_proc()**: Checks if the method returns a valid Proc object when datastore['HttpTrace'] is set to true.
2. **Rex::Proto::Http::Client#httptrace_use_callback()**: Executes the proc and checks if the output of the proc matches the expected output for a sample request/response. Also checks that a proper error message is displayed if the response received is nil.

### CODE REVIEW

File: lib/metasploit/framework/login_scanner/http.rb [hunk 1]
```diff
@@ -348,6 +324,38 @@ def attempt_login(credential)
           Result.new(result_opts)
         end
 
+
+        # This method returns a proc to log HTTP requests and responses
+        # if datastore['HttpTrace'] is set.
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/login_scanner/http.rb]: Just something I noticed but we are missing RDoc information for some of these functions and I think we should be adding these in considering some of the methods here appear to have RDoc documentation and other's don't. Particularly for these new methods of your own that you are adding in, you should be looking to add some RDoc headers in given you have the info to do this :)

File: lib/metasploit/framework/login_scanner/http.rb [hunk 2]
```diff
@@ -348,6 +324,38 @@ def attempt_login(credential)
...
+              if response
+                response = response.to_terminal_output(headers_only: http_trace_headers_only)
+                print_line("%clr#{response_color}#{response}%clr")
+              else
+                print_line('No response received')
+              end
+            }
+          end
+
+          proc_httptrace
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/login_scanner/http.rb:370]: This will be undefined if `http_trace` is not set.

File: lib/msf/core/exploit/remote/http_client.rb [hunk 3]
```diff
@@ -286,6 +264,38 @@ def connect_ws(opts={}, timeout = 20)
...
+        if response
+          response = response.to_terminal_output(headers_only: datastore['HttpTraceHeadersOnly'])
+          print_line("%clr#{response_color}#{response}%clr")
+        else
+          print_line('No response received')
+        end
+      }
+    end
+
+   proc_httptrace
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]: This will be undefined if `datastore['HttpTrace']` is a false value such as `nil`.

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 4]
```diff
@@ -30,4 +30,16 @@
...
   end
 
+  describe '#set_http_trace_proc' do
+    it 'configures http tracing' do
+      expect(subject.set_http_trace_proc(true)).to be_kind_of(Proc)
+    end
+  end
+
+  describe '#set_http_trace_proc' do
+    it 'does not log http tracing' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: These descriptions are too vauge. Configures HTTP tracing when what occurs? Does not log HTTP tracing when what occurs? Need to be more specific, otherwise its going to be harder for readers of RSpec output to determine what exactly went wrong.

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 5]
```diff
@@ -30,4 +30,16 @@
...
+      expect(subject.set_http_trace_proc(false)).to be_nil
+    end
+  end
+
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb:108]: These tests can probably also call the proc and check the output to be sure if we confident that output won't change.

File: spec/lib/rex/proto/http/client_spec.rb [hunk 6]
```diff
@@ -742,4 +742,44 @@ def excuse_needs_auth
...
+
+    it 'should execute the proc when defined' do
+      cli = Rex::Proto::Http::Client.new('127.0.0.1', http_trace_proc: http_trace_proc)
+      expect {cli.httptrace_use_callback(sample_request, sample_response)}.to output(expected_output).to_stdout
+    end
+
+    it 'should should give ideal message for nil response' do
+      cli = Rex::Proto::Http::Client.new('127.0.0.1', http_trace_proc: http_trace_proc)
+      expect {cli.httptrace_use_callback(sample_request, nil)}.to output(nil_response_output).to_stdout
+    end
```
Comment (gwillcox-r7) [inline spec/lib/rex/proto/http/client_spec.rb]: None of these tests are testing the new `set_http_trace_proc` method you added in? I'd also argue some of these tests could be added to your other spec as duplicates since you'll also want to test similar logic over there.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

**Comment by gwillcox-r7** [comment]
Note that `lib/msf/core/exploit/remote/http_client.rb` presently doesn't have any specs, despite being a core part of our code. Therefore we will be skipping adding the spec check to make sure we check the newly added `send_http_trace_proc` method since writing a proper spec file for this code would be quite complex and beyond the reasonable expectations of a medium sized GSOC project that is this late into its development cycle.

File: spec/lib/rex/proto/http/client_spec.rb [hunk 7]
```diff
@@ -742,4 +742,44 @@ def excuse_needs_auth
...
+    let(:nil_response_output) {
+      "####################\n# Request : \n####################\nGET / HTTP/1.1\nHost: www.google.com\n####################\n# Response : \n####################\nNo response received\n"
+    } 
+
+    it 'should execute the proc when defined' do
+      cli = Rex::Proto::Http::Client.new('127.0.0.1', http_trace_proc: http_trace_proc)
+      expect {cli.httptrace_use_callback(sample_request, sample_response)}.to output(expected_output).to_stdout
+    end
+
+    it 'should give ideal message for nil response' do
```
Comment (gwillcox-r7) [inline spec/lib/rex/proto/http/client_spec.rb]: What is an ideal message here? This is too vauge.

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 8]
```diff
@@ -31,13 +31,34 @@
...
       expect(subject.set_http_trace_proc(true)).to be_kind_of(Proc)
     end
+
+    #it 'should execute the proc when defined' do
+    #  expect { subject.set_http_trace_proc(true).call(sample_request, sample_response) }.to output(expected_output).to_stdout
+    #end
+
+    #it 'should give ideal message for nil response' do
+    #  expect { subject.set_http_trace_proc(true).call(sample_request, nil) }.to output(nil_response_output).to_stdout
+    #end
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: Where these meant to be commented out?

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 9]
```diff
@@ -31,13 +31,34 @@
...
+    let(:expected_output) {
+      "####################\n# Request : \n####################\nGET / HTTP/1.1\nHost: www.google.com\n####################\n# Response : \n####################\nHTTP/1.1 302 Found\nLocation: https://www.google.com/?gws_rd=ssl\n"
+    }
+    let(:nil_response_output) {
+      "####################\n# Request : \n####################\nGET / HTTP/1.1\nHost: www.google.com\n####################\n# Response : \n####################\nNo response received\n"
+    }
+
+    it 'returns a proc object when HttpTrace is set to true' do
       expect(subject.set_http_trace_proc(true)).to be_kind_of(Proc)
     end
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: This should be moved down to your `describe '#set_http_trace_proc' do` part as its related to the tests located there.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 10]
```diff
@@ -1,8 +1,33 @@
 
 require 'spec_helper'
 require 'metasploit/framework/login_scanner/http'
+require 'rex'
 
 RSpec.describe Metasploit::Framework::LoginScanner::HTTP do
+  include_context 'Msf::UIDriver'
```
Comment (adfoster-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb:5]:
```suggestion
  include_context 'Msf::DBManager'
  include_context 'Msf::UIDriver'
```

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 11]
```diff
@@ -1,8 +1,33 @@
...
+    def to_s(headers_only: false)
+      output_packet(false, headers_only: headers_only)
+    end
+  end
+
+  class Rex::Ui::Text::Output::Stdio
+    def support_color?
+      return false
+    end
+  end
```
Comment (adfoster-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: We should be able to delete this, if it was added to mock printing

```suggestion
```

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 12]
```diff
@@ -108,17 +107,33 @@
...
+
     it 'returns a proc object when HttpTrace is set to true' do
       expect(subject.set_http_trace_proc(true, false, nil)).to be_kind_of(Proc)
     end
 
     it 'should execute the proc when defined' do
       subject.set_http_trace_proc(true, false, nil).call(sample_request, sample_response)
       expect(@output).to eq expected_output
-    end
+    end
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: Extra space after end line?

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 13]
```diff
@@ -30,4 +35,160 @@
...
+
+    it 'should only log HTTP request in color when only one colour is specified' do
+      subject.set_http_trace_proc(true, false, 'yel/').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_single_color_output
+    end
+
+    it 'should only log HTTP request in color when only one color is specified without "/"' do
+      subject.set_http_trace_proc(true, false, 'yel').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_single_color_output
+    end
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: What if the option starts with a slash and then a color. So something like `/yel`?

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 14]
```diff
@@ -30,4 +35,160 @@
...
+      "GET / HTTP/1.1\nHost: www.google.com"
+    }
+
+    let(:sample_response) {
+      res = Rex::Proto::Http::Response.new(302,'Found')
+      allow(res).to receive(:body).and_return("Location: https://www.google.com/?gws_rd=ssl")
+      res
+    }
+
+    let(:expected_output) {
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: I think this could be renamed as this isn't very descriptive. What is expected? That could depend on the context. A better name would be something about it being a full request and response.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 15]
```diff
@@ -138,6 +138,40 @@
       ]
     }
 
+    let(:http_trace_single_color_output_leading_slash) {
+      [
+        "####################",
+        "# Request:",
+        "####################",
+        "%clr%bld%GET / HTTP/1.1",
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: Stray `%bld%` here?

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

**Review by gwillcox-r7 — APPROVED** [review]

File: lib/msf/core/exploit/remote/http_client.rb [hunk 16]
```diff
@@ -285,6 +263,42 @@ def connect_ws(opts={}, timeout = 20)
...
+  # @param http_trace [Bool] A Boolean representing the datastore['HttpTrace']
+  # to check if HttpTrace is set or unset.
+  #
+  # @return [Proc] A Proc object to log HTTP requests and responses, or nil if datastore['HttpTrace'] is unset
+  def set_http_trace_proc(http_trace)
+    proc_httptrace = nil
+    if http_trace
+      proc_httptrace = proc { |request, response| 
+        request_color, response_color = 
+          (datastore['HttpTraceColors'] || 'red/blu').split('/').map { |color| "%bld%#{color}" }
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]: Think this is still subject to the `%bld` issue that we mentioned earlier?

File: lib/msf/core/exploit/remote/http_client.rb [hunk 17]
```diff
@@ -285,6 +263,42 @@ def connect_ws(opts={}, timeout = 20)
...
+        
+        request = request.to_s(headers_only: datastore['HttpTraceHeadersOnly'])
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]: Why is this being done here and not in your [lib/metasploit/framework/login_scanner/http.rb](https://github.com/3V3RYONE/metasploit-framework/pull/7/files#diff-b00cbc1b487d443e3b3fe256e2489c8fe2c98ddd5d84a5c05d31bb9cb4f64633) definition?

Comment (gwillcox-r7) [reply lib/msf/core/exploit/remote/http_client.rb]: Also on a similar note shouldn't this be calling `to_terminal_output`? Why is `.to_s` being used here?

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: lib/msf/core/exploit/remote/http_client.rb (hunk 17 above)
Comment (gwillcox-r7) [reply lib/msf/core/exploit/remote/http_client.rb]: https://github.com/rapid7/metasploit-framework/search?q=to_terminal_output shows that only the packet response has `to_terminal_output` call on it to convert from a packet to proper terminal output. Therefore these changes are in line with that current code pattern, as was mentioned by @3V3RYONE on the call this morning.

File: lib/metasploit/framework/login_scanner/http.rb [hunk 18]
```diff
@@ -246,32 +247,8 @@ def send_request(opts)
           context         = opts['context'] || { 'Msf' => framework, 'MsfExploit' => framework_module}
 
           res = nil
+          http_trace_proc = set_http_trace_proc(http_trace, http_trace_headers_only, http_trace_colors)
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/login_scanner/http.rb:249]: This relies on `http_trace` definition which is inside of [lib/msf/core/exploit/remote/http_client.rb](https://github.com/3V3RYONE/metasploit-framework/pull/7/files#diff-181d5970bce86f6a48bb550ed0dbade6ea0d946645995e0ff7884e85fe3fe91a), yet we don't pull that in.

If this library is loaded without [lib/msf/core/exploit/remote/http_client.rb](https://github.com/3V3RYONE/metasploit-framework/pull/7/files#diff-181d5970bce86f6a48bb550ed0dbade6ea0d946645995e0ff7884e85fe3fe91a) also being loaded, then it will fail to find the definition for this function.

I imagine we may want to require the other library which will prevent duplicates from being loaded whilst still ensuring the definition we need here is in memory.

File: lib/metasploit/framework/login_scanner/http.rb [hunk 19]
```diff
@@ -348,6 +325,44 @@ def attempt_login(credential)
           Result.new(result_opts)
         end
 
+
+        # Defines a proc to log HTTP requests and responses
+        #
+        # @param http_trace [Bool] A Boolean representing the datastore['HttpTrace'] to check if HttpTrace is set or unset.
+        # @param http_trace_headers_only [Bool] A Boolean representing the datastore['HttpTraceHeadersOnly'] to check if only HTTP headers need to be logged.
+        # @param http_trace_colors [String] A string representing the datastore['HttpTraceColors'] to specify the colors in which request and response need to be logged.
+        # @return [Proc] A Proc object to log HTTP requests and responses, or nil if the datastore['HttpTrace'] is unset
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/login_scanner/http.rb]:
```suggestion
        # @param http_trace [Bool] A Boolean representing datastore['HttpTrace'] to check if HttpTrace is set or unset.
        # @param http_trace_headers_only [Bool] A Boolean representing datastore['HttpTraceHeadersOnly'] to check if only HTTP headers need to be logged.
        # @param http_trace_colors [String] A string representing datastore['HttpTraceColors'] to specify the colors in which request and response need to be logged.
        # @return [Proc] A Proc object to log HTTP requests and responses, or nil if datastore['HttpTrace'] is unset
```

File: lib/metasploit/framework/login_scanner/http.rb [hunk 20]
```diff
@@ -348,6 +325,44 @@ def attempt_login(credential)
...
+        def set_http_trace_proc(http_trace, http_trace_headers_only, http_trace_colors)
+          proc_httptrace = nil
+          if http_trace
+            proc_httptrace = proc { |request, response|
+              request_color, response_color =
+                (http_trace_colors || 'red/blu').split('/').map { |color| color.blank? ? "" : "%bld%#{color}" }
```
Comment (gwillcox-r7) [inline lib/metasploit/framework/login_scanner/http.rb]: This isn't actually working as intended in my tests:

```
>> a = "red/blue"
>> a.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
>> a
>> b = ""
>> c = "red/"
=> "red/"
>> d = "/red"
=> "/red"
>> a.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
=> ["%bldred", "%bldblue"]
>> b.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
=> []
>> c.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
=> ["%bldred"]
>> d.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
=> ["", "%bldred"]
>> b[0]
=> nil
>> b[1]
=> nil
>> d[0]
=> "/"
>> request_color, response_color = c.split('/').map { |color| color.blank? ? "" : "%bld#{color}" }
=> ["%bldred"]
>> request_color
=> "%bldred"
>> response_color
=> nil
>> 
```
You are being saved here by the fact that `nil.to_s` converts to the same thing as `""` but anyone looking at your code would assume the value should be `""` when in reality this is not the case.

Comment (gwillcox-r7) [reply lib/metasploit/framework/login_scanner/http.rb]: And I think I figured out why this isn't working:

```
>> "red/".split('/')
=> ["red"]
>> 
```
Looks like we are only getting one element if there is a slash.

Here is a solution I came up with:

```
>> "".gsub('/', ' / ').split('/').map { |color| color&.strip!.blank? ? "" : "%bld#{color}" }
=> []
>> "/".gsub('/', ' / ').split('/').map { |color| color&.strip!.blank? ? "" : "%bld#{color}" }
=> ["", ""]
>> "red/".gsub('/', ' / ').split('/').map { |color| color&.strip!.blank? ? "" : "%bld#{color}" }
=> ["%bldred", ""]
>> "/red".gsub('/', ' / ').split('/').map { |color| color&.strip!.blank? ? "" : "%bld#{color}" }
=> ["", "%bldred"]
>> 
```

Only issue would be if the entry is blank, in which case we can just change it to be `"/"` and this will solve the issue with the new code.

File: lib/msf/core/exploit/remote/http_client.rb [hunk 21]
```diff
@@ -150,30 +150,8 @@ def connect(opts={})
...
-        if response
-          response = response.to_terminal_output(headers_only: datastore['HttpTraceHeadersOnly'])
-          print_line("%clr#{response_color}#{response}%clr")
-        else
-          print_line('No response received')
-        end
-      }
-    end
+    
+    http_trace_proc = set_http_trace_proc(datastore['HttpTrace'], datastore['HttpTraceHeadersOnly'], datastore['HttpTraceColors'])
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]:
```suggestion
    http_trace_proc = set_http_trace_proc(http_trace, http_trace_headers_only, http_trace_colors)
```
You are already defining these accessor methods within your class yet you aren't using them here and you are using them in the other library? Seems a bit odd :) This will also make this line shorter and easier to read.

File: lib/msf/core/exploit/remote/http_client.rb [hunk 22]
```diff
@@ -285,6 +263,43 @@ def connect_ws(opts={}, timeout = 20)
...
     socket.extend(Rex::Proto::Http::WebSocket::Interface)
   end
+  
+  
+  # Defines a proc to log HTTP requests and responses
+  #
+  # @param http_trace [Bool] A Boolean representing the datastore['HttpTrace'] to check if HttpTrace is set or unset.
+  # @param http_trace_headers_only [Bool] A Boolean representing the datastore['HttpTraceHeadersOnly'] to check if only HTTP headers need to be logged.
+  # @param http_trace_colors [String] A string representing the datastore['HttpTraceColors'] to specify the colors in which request and response need to be logged.
+  # @return [Proc] A Proc object to log HTTP requests and responses, or nil if the datastore['HttpTrace'] is unset
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]:
```suggestion
  # @param http_trace [Bool] A Boolean representing datastore['HttpTrace'] to check if HttpTrace is set or unset.
  # @param http_trace_headers_only [Bool] A Boolean representing datastore['HttpTraceHeadersOnly'] to check if only HTTP headers need to be logged.
  # @param http_trace_colors [String] A string representing datastore['HttpTraceColors'] to specify the colors in which request and response need to be logged.
  # @return [Proc] A Proc object to log HTTP requests and responses, or nil if datastore['HttpTrace'] is unset
```

File: lib/msf/core/exploit/remote/http_client.rb [hunk 23]
```diff
@@ -285,6 +263,43 @@ def connect_ws(opts={}, timeout = 20)
...
+  def set_http_trace_proc(http_trace, http_trace_headers_only, http_trace_colors)
+    proc_httptrace = nil
+    if http_trace
+      proc_httptrace = proc { |request, response| 
+        request_color, response_color = 
+          (http_trace_colors || 'red/blu').split('/').map { |color| color.blank? ? "" : "%bld%#{color}" }
```
Comment (gwillcox-r7) [inline lib/msf/core/exploit/remote/http_client.rb]: Comment above on handling of code here also applies.

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 24]
```diff
@@ -30,4 +35,213 @@
...
+        "\r",
+        "Location: https://www.google.com/?gws_rd=ssl%clr"
+      ]
+    }
+
+    it 'returns a proc object when HttpTrace is set to true' do
+      expect(subject.set_http_trace_proc(true, false, nil)).to be_kind_of(Proc)
+    end
+
+    it 'should execute the proc when defined' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]: Don't think this is accurate description here. This should probably reflect the fact that we have are sending it a normal request and response and that we are expecting it to return the normal full output.

The proc could be a `nil` object due to the HTTPTrace flag not being set and it would still be "defined", just as a `nil` object and not as a `proc` object.

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 25]
```diff
@@ -30,4 +35,213 @@
...
+      subject.set_http_trace_proc(true, true, nil).call(sample_request, sample_response)
+      expect(@output).to eq headers_only_output
+    end
+
+    it 'should log HTTP requests and responses with body when HttpTraceHeadersOnly is unset' do
+      subject.set_http_trace_proc(true, nil, nil).call(sample_request, sample_response)
+      expect(@output).to eq normal_request_response_output
+    end
+
+    it 'should log HTTP requests and responses in the specified color' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]:
```suggestion
    it 'should log HTTP requests and responses in the respective colors specified' do
```

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 26]
```diff
@@ -30,4 +35,213 @@
...
+      subject.set_http_trace_proc(true, false, 'blu/grn').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_colors_output
+    end
+
+    it 'should only log HTTP request in color when only one colour is specified' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]:
```suggestion
    it 'should only log HTTP request in color when only one color is specified followed by a trailing "/"' do
```

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 27]
```diff
@@ -30,4 +35,213 @@
...
+      subject.set_http_trace_proc(true, false, 'yel/').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_single_color_output
+    end
+
+    it 'should only log HTTP response in color when only one colour is specified after a leading "/"' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]:
```suggestion
    it 'should only log HTTP response in color when only one color is specified after a leading "/"' do
```

File: spec/lib/metasploit/framework/login_scanner/http_spec.rb [hunk 28]
```diff
@@ -30,4 +35,213 @@
...
+      subject.set_http_trace_proc(true, false, '/yel').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_single_color_output_leading_slash
+    end
+
+    it 'should not log HTTP request and response in color when HttpTraceColors is set to "/"' do
+      subject.set_http_trace_proc(true, false, '/').call(sample_request, sample_response)
+      expect(@output).to eq http_trace_no_color_output
+    end
+
+    it 'should only log HTTP request in color when only one color is specified without "/"' do
```
Comment (gwillcox-r7) [inline spec/lib/metasploit/framework/login_scanner/http_spec.rb]:
```suggestion
    it 'should only log HTTP request in color when only one color is specified without any "/"s' do
```

File: spec/lib/rex/proto/http/client_spec.rb [hunk 29]
```diff
@@ -742,4 +742,44 @@ def excuse_needs_auth
...
+    let(:sample_request) {
+      "GET / HTTP/1.1\nHost: www.google.com"
+    }
+    let(:sample_response) {
+      "HTTP/1.1 302 Found\nLocation: https://www.google.com/?gws_rd=ssl"
+    }
+    let(:http_trace_proc) { 
+      proc { |request, response|
+        puts("#" * 20)
+        puts("# Request : ")
```
Comment (gwillcox-r7) [inline spec/lib/rex/proto/http/client_spec.rb]:
```suggestion
        puts("# Request:")
```

Applies for the rest, just looks a bit untidy to have trailing spaces and whitespace like this. Additionally this new format better matches the existing proc format for printing output that we have in our current HTTPTrace implementation.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

**Comment by gwillcox-r7** [comment]
@3V3RYONE left some comments and was going to push up the associated fixes but looks like I don't have permissions to push to this branch on your repository. Let me know if you have any questions.

File: lib/metasploit/framework/login_scanner/http.rb (hunk 20 above)
Comment (3V3RYONE) [reply lib/metasploit/framework/login_scanner/http.rb]: Thanks for the suggestion. Accounts to normalise for maximum cases. But in the case of specifying a single color without any `/` the tests fail.

For instance, 
```
>> "red".gsub('/', ' / ').split('/').map { |color| color&.strip!.blank? ? "" : "%bld#{color}" }
=> [""]
```
Ideally when we provide one color as input, the request should be logged in the specified color. But with this change, the request and response are logged without any color.

File: lib/metasploit/framework/login_scanner/http.rb (hunk 18 above)
Comment (gwillcox-r7) [reply lib/metasploit/framework/login_scanner/http.rb:249]: Annnnnd this is crossing things over a bit since now we are crossing namespaces in the wrong direction 🤦

File: lib/metasploit/framework/login_scanner/http.rb (hunk 20 above)
Comment (gwillcox-r7) [reply lib/metasploit/framework/login_scanner/http.rb]: Oh okay so this may also be cause I used .strip! instead of .strip which means it will return `nil` and then that is passed into `blank?` which was causing issues. Also we were missing `%` before `#{color}` which explains the colorless output.

File: lib/metasploit/framework/login_scanner/http.rb (hunk 18 above)
Comment (gwillcox-r7) [reply lib/metasploit/framework/login_scanner/http.rb:249]: Going to close this since even in RSpec tests this worked out fine so not sure its needed. Including it creates more issues as well so honestly don't think its worth including this line given the combination of these two results.

File: lib/metasploit/framework/login_scanner/http.rb (hunk 20 above)
Comment (gwillcox-r7) [reply lib/metasploit/framework/login_scanner/http.rb]: Fixed with https://github.com/3V3RYONE/metasploit-framework/pull/7/commits/29c8adbb67dbaae67bf1ce040ca3bf852a54465c

**Comment by gwillcox-r7** [comment]
@3V3RYONE So turns out I used `.strip!` instead of `.strip` in the code which meant it was return a `nil` object to `blank?` which was causing issues. Also we were missing `%` before `#{color}` which explains the colorless output. With these issues fixed the specs should now be working.

I also went ahead and update the documentation on the way the code I wrote works since it is a little unusual 👍

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by 3V3RYONE** [comment]
Thanks a lot @gwillcox-r7 , I figured out the error was in `strip!` and was working on that in a local commit. All the other suggestions were fixed in my local commit. Thanks for chiming in and helping a lot :)
